In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
data = Path('Resources/lending_data.csv')
df = pd.read_csv(data)

# Review the DataFrame
# YOUR CODE HERE!
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y = df["loan_status"]

# Separate the X variable, the features
# YOUR CODE HERE!
X = df.drop(columns="loan_status")

In [4]:
# Review the y variable Series
# YOUR CODE HERE!
y

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64

In [5]:
# Review the X variable DataFrame
# YOUR CODE HERE!
X

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [15]:
# Check the balance of our target values
# YOUR CODE HERE!
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [16]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [18]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
classifier = LogisticRegression(solver='lbfgs', random_state=1)
# Fit the model using training data
# YOUR CODE HERE!
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [19]:
# Make a prediction using the testing data
# YOUR CODE HERE!
predictions = classifier.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [20]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!
balanced_accuracy_score(y_test, predictions)

0.9520479254722232

In [21]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
confusion_matrix(y_test, predictions)

array([[18663,   102],
       [   56,   563]], dtype=int64)

In [23]:
# Print the classification report for the model
# YOUR CODE HERE!
target_names = ["healthy loan", "high-risk loan"]
print(classification_report(y_test, predictions, target_names=target_names))

                precision    recall  f1-score   support

  healthy loan       1.00      0.99      1.00     18765
high-risk loan       0.85      0.91      0.88       619

      accuracy                           0.99     19384
     macro avg       0.92      0.95      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model performed well in predicting healthy and high-risk labels, as indicated by the classification report. Both classes achieved high precision and recall. For the healthy loan class, a precision of 1.00 and recall of 0.99 were achieved, while the high-risk loan class obtained a precision of 0.85 and recall of 0.91. Overall, the model achieved an accuracy of 0.99, indicating strong performance in predicting both healthy and high-risk loans.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [24]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
sample = RandomOverSampler(random_state=1)
# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!
X_resampled, y_resampled = sample.fit_resample(X_train, y_train)

In [25]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!
y_resampled.value_counts()

1    56271
0    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [26]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
classifier2 = LogisticRegression(solver='lbfgs', random_state=1)
# Fit the model using the resampled training data
# YOUR CODE HERE!
classifier2.fit(X_resampled, y_resampled)
# Make a prediction using the testing data
# YOUR CODE HERE!
predictions2 = classifier2.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [27]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
balanced_accuracy_score(y_test, predictions2)

0.9936781215845847

In [28]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
confusion_matrix(y_test, predictions2)

array([[18649,   116],
       [    4,   615]], dtype=int64)

In [29]:
# Print the classification report for the model
# YOUR CODE HERE!
target_names = ["healthy loan", "high-risk loan"]
print(classification_report(y_test, predictions2, target_names=target_names))

                precision    recall  f1-score   support

  healthy loan       1.00      0.99      1.00     18765
high-risk loan       0.84      0.99      0.91       619

      accuracy                           0.99     19384
     macro avg       0.92      0.99      0.95     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model fit with oversampled data performed well in predicting healthy and high-risk labels, as indicated by the classification report. Both classes achieved high precision and recall. For the healthy loan class, a precision of 1.00 and recall of 0.99 were achieved, while the high-risk loan class obtained a precision of 0.84 and recall of 0.99. Overall, the model achieved an accuracy of 0.99, indicating strong performance in predicting both healthy and high-risk loans.